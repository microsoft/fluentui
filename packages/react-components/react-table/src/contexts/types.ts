export type TableContextValue = {
  size: 'small' | 'smaller' | 'medium';

  noNativeElements: boolean;
};
