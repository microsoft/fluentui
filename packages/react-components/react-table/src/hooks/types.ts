import * as React from 'react';
import { SortDirection } from '../components/Table/Table.types';
import { TableHeaderCellProps } from '../components/TableHeaderCell/TableHeaderCell.types';

export type TableRowId = string | number;
export type TableColumnId = string | number;
export type SelectionMode = 'single' | 'multiselect';

export interface SortState {
  sortColumn: TableColumnId | undefined;
  sortDirection: SortDirection;
}

export interface OnSelectionChangeData {
  selectedItems: Set<TableRowId>;
}

export interface CreateTableColumnOptions<TItem> extends Partial<TableColumnDefinition<TItem>> {
  columnId: TableColumnId;
}

export interface TableColumnDefinition<TItem> {
  columnId: TableColumnId;
  compare: (a: TItem, b: TItem) => number;
  renderHeaderCell: () => React.ReactNode;
  renderCell: (item: TItem) => React.ReactNode;
}

export type RowEnhancer<TItem, TRowState extends TableRowData<TItem> = TableRowData<TItem>> = (
  row: TableRowData<TItem>,
) => TRowState;

export interface TableSortState<TItem> {
  /**
   * Current sort direction
   */
  sortDirection: SortDirection;
  /**
   * Column id of the currently sorted column
   */
  sortColumn: TableColumnId | undefined;
  /**
   * Set the sort direction for the specified column
   */
  setColumnSort: (event: React.SyntheticEvent, columnId: TableColumnId, sortDirection: SortDirection) => void;
  /**
   * Toggles the sort direction for specified column
   */
  toggleColumnSort: (event: React.SyntheticEvent, columnId: TableColumnId) => void;
  /**
   * Returns the sort direction if a column is sorted,
   * returns undefined if the column is not sorted
   */
  getSortDirection: (columnId: TableColumnId) => SortDirection | undefined;

  /**
   * Sorts rows and returns a **shallow** copy of original items
   */
  sort: <TRowState extends TableRowData<TItem>>(rows: TRowState[]) => TRowState[];
}

export interface TableSelectionState {
  /**
   * Clears all selected rows
   */
  clearRows: (e: React.SyntheticEvent) => void;
  /**
   * Selects single row
   */
  selectRow: (e: React.SyntheticEvent, rowId: TableRowId) => void;
  /**
   * De-selects single row
   */
  deselectRow: (e: React.SyntheticEvent, rowId: TableRowId) => void;
  /**
   * Toggle selection of all rows
   */
  toggleAllRows: (e: React.SyntheticEvent) => void;
  /**
   * Toggle selection of single row
   */
  toggleRow: (e: React.SyntheticEvent, rowId: TableRowId) => void;
  /**
   * Collection of row ids corresponding to selected rows
   */
  selectedRows: Set<TableRowId>;
  /**
   * Whether all rows are selected
   */
  allRowsSelected: boolean;
  /**
   * Whether some rows are selected
   */
  someRowsSelected: boolean;

  /**
   * Checks if a given rowId is selected
   */
  isRowSelected: (rowId: TableRowId) => boolean;

  selectionMode: SelectionMode;
}

export interface TableRowData<TItem> {
  /**
   * User provided data
   */
  item: TItem;
  /**
   * The row id, defaults to index position in the collection
   */
  rowId: TableRowId;
}

export interface TableFeaturesState<TItem> extends Pick<UseTableFeaturesOptions<TItem>, 'items' | 'getRowId'> {
  /**
   * The row data for rendering
   * @param rowEnhancer - Enhances the row with extra user data
   */
  getRows: <TRowState extends TableRowData<TItem> = TableRowData<TItem>>(
    rowEnhancer?: RowEnhancer<TItem, TRowState>,
  ) => TRowState[];
  /**
   * State and actions to manage row selection
   */
  selection: TableSelectionState;
  /**
   * State and actions to manage row sorting
   */
  sort: TableSortState<TItem>;
  /**
   * Table columns
   */
  columns: TableColumnDefinition<TItem>[];
  /**
   * State and actions to manage column resizing
   */
  // eslint-disable-next-line @typescript-eslint/naming-convention
  columnSizing_unstable: TableColumnSizingState;
  /**
   * A React.Ref object to be set as a ref for the table.
   * Used with column resizing.
   */
  tableRef: React.Ref<HTMLDivElement>;
}

export interface UseTableSortOptions {
  /**
   * Used to control sorting
   */
  sortState?: SortState;
  /**
   * Used in uncontrolled mode to set initial sort column and direction on mount
   */
  defaultSortState?: SortState;
  /**
   * Called when sort changes
   */
  onSortChange?(e: React.SyntheticEvent, state: SortState): void;
}

export interface UseTableSelectionOptions {
  /**
   * Can be multi or single select
   */
  selectionMode: SelectionMode;
  /**
   * Used in uncontrolled mode to set initial selected rows on mount
   */
  defaultSelectedItems?: Iterable<TableRowId>;
  /**
   * Used to control row selection
   */
  selectedItems?: Iterable<TableRowId>;
  /**
   * Called when selection changes
   */
  onSelectionChange?(e: React.SyntheticEvent, data: OnSelectionChangeData): void;
}

export interface UseTableFeaturesOptions<TItem> {
  columns: TableColumnDefinition<TItem>[];
  items: TItem[];
  getRowId?: (item: TItem) => TableRowId;
}

export type TableFeaturePlugin = <TItem>(tableState: TableFeaturesState<TItem>) => TableFeaturesState<TItem>;

export interface ColumnWidthState {
  columnId: TableColumnId;
  width: number;
  minWidth: number;
  idealWidth: number;
  padding: number;
}

export type ColumnSizingTableHeaderCellProps = Pick<TableHeaderCellProps, 'style' | 'aside'>;
export type ColumnSizingTableCellProps = Pick<TableHeaderCellProps, 'style'>;

export interface TableColumnSizingState {
  getOnMouseDown: (columnId: TableColumnId) => (e: React.MouseEvent | React.TouchEvent) => void;
  setColumnWidth: (columnId: TableColumnId, newSize: number) => void;
  getColumnWidths: () => ColumnWidthState[];
  getTableHeaderCellProps: (columnId: TableColumnId) => ColumnSizingTableHeaderCellProps;
  getTableCellProps: (columnId: TableColumnId) => ColumnSizingTableCellProps;
}

export type ColumnResizeState = {
  getColumnWidth: (columnId: TableColumnId) => number;
  setColumnWidth: (e: TouchEvent | MouseEvent | undefined, data: { columnId: TableColumnId; width: number }) => void;
  getColumnById: (columnId: TableColumnId) => ColumnWidthState | undefined;
  getColumns: () => ColumnWidthState[];
};

export type TableColumnSizingOptions = Record<
  TableColumnId,
  Partial<Pick<ColumnWidthState, 'minWidth' | 'idealWidth' | 'padding'>> & { defaultWidth?: number }
>;

export type UseTableColumnSizingParams = {
  columnSizingOptions?: TableColumnSizingOptions;
  onColumnResize?: (e: TouchEvent | MouseEvent | undefined, data: { columnId: TableColumnId; width: number }) => void;
  containerWidthOffset?: number;
};
