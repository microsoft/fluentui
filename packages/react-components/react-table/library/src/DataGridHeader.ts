export * from './components/DataGridHeader/index';
