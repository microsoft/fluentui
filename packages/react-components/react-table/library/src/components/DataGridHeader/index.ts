export { DataGridHeader } from './DataGridHeader';
export type { DataGridHeaderProps, DataGridHeaderSlots, DataGridHeaderState } from './DataGridHeader.types';
export { renderDataGridHeader_unstable } from './renderDataGridHeader';
export { useDataGridHeader_unstable } from './useDataGridHeader';
export { dataGridHeaderClassNames, useDataGridHeaderStyles_unstable } from './useDataGridHeaderStyles.styles';
