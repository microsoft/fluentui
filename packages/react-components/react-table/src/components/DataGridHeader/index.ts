export * from './DataGridHeader';
export * from './DataGridHeader.types';
export * from './renderDataGridHeader';
export * from './useDataGridHeader';
export * from './useDataGridHeaderStyles.styles';
