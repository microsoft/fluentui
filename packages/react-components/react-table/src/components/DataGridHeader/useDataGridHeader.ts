import * as React from 'react';
import type { DataGridHeaderProps, DataGridHeaderState } from './DataGridHeader.types';
import { useTableHeader_unstable } from '../TableHeader/useTableHeader';

/**
 * Create the state required to render DataGridHeader.
 *
 * The returned state can be modified with hooks such as useDataGridHeaderStyles_unstable,
 * before being passed to renderDataGridHeader_unstable.
 *
 * @param props - props from this instance of DataGridHeader
 * @param ref - reference to root HTMLElement of DataGridHeader
 */
export const useDataGridHeader_unstable = (
  props: DataGridHeaderProps,
  ref: React.Ref<HTMLElement>,
): DataGridHeaderState => {
  return useTableHeader_unstable({ ...props, as: 'div' }, ref);
};
