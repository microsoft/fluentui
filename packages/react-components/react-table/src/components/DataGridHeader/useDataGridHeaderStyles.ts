import { mergeClasses } from '@griffel/react';
import type { DataGridHeaderSlots, DataGridHeaderState } from './DataGridHeader.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { useTableHeaderStyles_unstable } from '../TableHeader/useTableHeaderStyles';

export const dataGridHeaderClassNames: SlotClassNames<DataGridHeaderSlots> = {
  root: 'fui-DataGridHeader',
};

/**
 * Apply styling to the DataGridHeader slots based on the state
 */
export const useDataGridHeaderStyles_unstable = (state: DataGridHeaderState): DataGridHeaderState => {
  useTableHeaderStyles_unstable(state);
  state.root.className = mergeClasses(dataGridHeaderClassNames.root, state.root.className);

  return state;
};
