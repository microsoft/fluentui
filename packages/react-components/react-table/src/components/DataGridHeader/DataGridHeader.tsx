import * as React from 'react';
import { useDataGridHeader_unstable } from './useDataGridHeader';
import { renderDataGridHeader_unstable } from './renderDataGridHeader';
import { useDataGridHeaderStyles_unstable } from './useDataGridHeaderStyles';
import type { DataGridHeaderProps } from './DataGridHeader.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * DataGridHeader component
 */
export const DataGridHeader: ForwardRefComponent<DataGridHeaderProps> = React.forwardRef((props, ref) => {
  const state = useDataGridHeader_unstable(props, ref);

  useDataGridHeaderStyles_unstable(state);
  return renderDataGridHeader_unstable(state);
});

DataGridHeader.displayName = 'DataGridHeader';
