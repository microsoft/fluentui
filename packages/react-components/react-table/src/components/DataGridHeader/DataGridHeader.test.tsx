import * as React from 'react';
import { render } from '@testing-library/react';
import { DataGridHeader } from './DataGridHeader';
import { isConformant } from '../../testing/isConformant';
import { DataGridHeaderProps } from './DataGridHeader.types';

describe('DataGridHeader', () => {
  isConformant<DataGridHeaderProps>({
    Component: DataGridHeader,
    displayName: 'DataGridHeader',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<DataGridHeader>Default DataGridHeader</DataGridHeader>);
    expect(result.container).toMatchSnapshot();
  });
});
