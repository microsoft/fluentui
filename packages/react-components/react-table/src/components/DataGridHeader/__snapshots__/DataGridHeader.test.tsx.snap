// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataGridHeader renders a default state 1`] = `
<div>
  <div
    class="fui-DataGridHeader fui-TableHeader"
    role="rowgroup"
  >
    Default DataGridHeader
  </div>
</div>
`;
